### PR TITLE
TARDIS "scientific attention" PR review helper skill

### PR DESCRIPTION
**Pull request by andrewfullard**
### :pencil: Description

**Type:**  :roller_coaster: `infrastructure`

This PR adds a skill that provides review support for TARDIS developers. It is intended to point them at relevant areas where scientific attention is most valuable. It is not likely to be fully comprehensive, and does not look closely at code style, bugs etc.

### :pushpin: Example review of #3365 

This brief highlights science questions for specialist review; it does not determine whether the proposed change is correct.

### What the change aims to do

[[PR #3365](https://github.com/tardis-sn/tardis/pull/3365)](https://github.com/tardis-sn/tardis/pull/3365) introduces one-dimensional homologous mesh objects and density solvers for uniform, power-law, exponential, and Branch85 W7 profiles. It assumes density is evaluated at shell-center velocity, \(r=vt\), and W7 follows

\[
\rho(v,t)=\rho_0\left(\frac{v}{v_0}\right)^{-7}\left(\frac{t_0}{t}\right)^3.
\]

The starting version already implements the \(v^{-7}\) profile followed by one \(t^{-3}\) scaling, so a reviewer should clarify what scientific error relative to that version the PR intends to correct.

#### 1. Homologous time dilution is applied twice — `close review`

**Where to look:** The [[W7 solver scales density to the explosion time](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L226-L250)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L226-L250), but the [[transition parser returns that result with the original reference time](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L60-L70)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L60-L70), after which [[model construction applies the same \(t^{-3}\) transformation again](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_composition_configuration.py#L46-L54)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_composition_configuration.py#L46-L54).

**Why it matters:** This code path directly produces an overall \(t^{-6}\), rather than \(t^{-3}\), dependence for configured W7 models. For the standard 13-day configuration with \(t_0=0.000231481\) day, the second scaling introduces an additional factor of \(5.65\times10^{-15}\). This would strongly suppress ejecta mass density, elemental and electron number densities, level populations, and optical depths.

**Questions to answer:**

- Should every solver return density at \(t_0\), leaving time evolution to model construction, or should it return density at \(t_\mathrm{explosion}\)? Only one layer should own the transformation.
- Does \(\rho(t_2)/\rho(t_1)=(t_1/t_2)^3\) and does \(\rho V\) remain constant for fixed velocity shells?
- Can the authors show a derivation or Branch85 reference establishing the intended formula? The schema calls it a “seventh-order polynomial fit,” while the implementation is a \(v^{-7}\) power law.

**What evidence exists:** The [[new isolated-solver test](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/tests/test_density_from_config.py#L17-L59)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/tests/test_density_from_config.py#L17-L59) contains useful first- and last-shell expected values, but bypasses the parser and therefore cannot detect the second scaling. The existing full-model test contains the same expected values, but did not run for this version.

#### 2. Shell-center velocities are reinterpreted as interfaces — `close review`

**Where to look:** Model construction first calculates shell centers, then the [[changed adapter prepends another zero and labels those centers as mesh interfaces](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L47-L62)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L47-L62). The [[solvers average adjacent values again](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L105-L124)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L105-L124).

**Why it matters:** This shifts the velocity at which every non-uniform analytic density is evaluated. In the paper configuration, the first retained shell changes from 11,225 to 8,362.5 km s\(^{-1}\). For a \(v^{-7}\) profile, that alone raises its density by a factor of about 7.85 before the time-scaling effect is included. Power-law and exponential configurations, including configured CSVY models, follow the same adapter.

The new mesh description also calls cell zero a central volume, although the new configuration parser begins at the positive configured inner velocity. That convention needs resolution before any later calculation masks or specially treats cell zero.

**Questions to answer:**

- Should the adapter receive the original \(N+1\) interfaces, or should solvers accept the already-computed \(N\) centers?
- Does each retained shell use exactly \((v_\mathrm{inner}+v_\mathrm{outer})/2\), including nonuniform grids and configured inner/outer boundaries?
- Are shell indexing and the synthetic central point identical between classic configuration, CSVY, and file-based models?

**What evidence exists:** Solver tests confirm midpoint evaluation when supplied genuine interfaces. The configuration tests construct genuine interfaces directly and therefore bypass the problematic center-to-interface conversion. There is no test of the complete adapter path on an irregular grid or with boundary cuts.

#### 3. Scientific validation did not execute — `check carefully`

**Where to look:** The [[W7 unit tests omit the now-required reference velocity](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/tests/test_density_solvers.py#L175-L214)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/tests/test_density_solvers.py#L175-L214), while the [[parser test expects the obsolete solver type](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/tests/test_density_parser.py#L60-L75)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/tests/test_density_parser.py#L60-L75).

**Why it matters:** These inconsistencies obstruct the evidence intended to validate the W7 formulation, even before the scientific assertions are reached.

**Questions to answer:**

- After repairing the tests, does the full configuration-to-composition path reproduce the existing W7 expected densities?
- Is there a two-epoch \(t^{-3}\) test, a shell-mass conservation test, and a comparison of resulting number densities or optical depths?
- Should stored reference spectra be compared after the density path is corrected?

**What evidence exists:** The exact-version [[unit-test run](https://github.com/tardis-sn/tardis/actions/runs/20195212322)](https://github.com/tardis-sn/tardis/actions/runs/20195212322) never started its tests because the data-cache prerequisite failed. [[Full tests](https://github.com/tardis-sn/tardis/actions/runs/20435912673)](https://github.com/tardis-sn/tardis/actions/runs/20435912673) and [[stored-reference comparisons](https://github.com/tardis-sn/tardis/actions/runs/29265627863)](https://github.com/tardis-sn/tardis/actions/runs/29265627863) were skipped. [[Benchmarks succeeded](https://github.com/tardis-sn/tardis/actions/runs/29265627885)](https://github.com/tardis-sn/tardis/actions/runs/29265627885), but measured execution time rather than densities, conservation, or spectra.

### How the effects move through the simulation

The two demonstrated transformations multiply together in the initial density field. Density is combined with mass fractions to form elemental and isotopic number densities; these set ion, electron, and level populations. Those populations determine electron-scattering and Sobolev optical depths, interaction probabilities, and ultimately plasma convergence and the emergent spectrum. No output-level comparison currently constrains these consequences.

### What has been checked

The complete three-dot change set, configuration defaults, density equations, mesh conventions, model-construction path, relevant tests, discussion, and exact-version automated records were inspected. No proposed code, tests, notebooks, or benchmarks were run locally.

### Version reviewed

- PR: [[#3365 — Fixes W7 density solver implementation and reorganizes tests](https://github.com/tardis-sn/tardis/pull/3365)](https://github.com/tardis-sn/tardis/pull/3365)
- Starting identifier reported by GitHub: `a1a284343ed3cba7acf5cd6606c0ec0e166b8e0c`
- Proposed identifier: `5a420c540f44bed66c98d3561258b298a47228a7`

The proposed identifier remained unchanged at final recheck. The starting snapshot is formally inconsistent: a fresh fetch of `master` produced `3caeb850d4421605c263e1de29aae392aca09fc3`, while GitHub continued to report `a1a2843…` after the required re-query. Both have the same common ancestor with the proposed version, so the complete PR-owned change set reviewed is unchanged.

### CODE REVIEW

File: .agents/skills/tardis-pr-review/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+   - Introduce a source link as **Where to look**, not a `line anchor`.
+
+3. When a technical software term is unavoidable, define it in one short
+   sentence at first use. Do not expect the reader to interpret branch names,
+   workflow names, test markers, or abbreviated check names.
+
+4. Focus each area on the scientific choice and its possible effect. Mention
+   code mechanics only when they help the reader understand that effect.
+
+## Report a Scientific-Review Brief
```
Comment (andrewfullard) [inline .agents/skills/tardis-pr-review/SKILL.md:210]: Brief is the operative word, keep it short, do not add spurious information or questions not directly motivated by the code

**Comment by andrewfullard** [comment]
Briefer version with latest commit:

This brief highlights science questions for specialist review; it does not determine whether the proposed change is correct. PR 3365 aims to replace the classic density parser with solver classes and revise the Branch85 W7 density path.

## Areas needing scientific attention

### W7 time scaling may be applied twice — close review

**Where to look**: [[W7 solver](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L192-L200)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L192-L200) and [[configuration parser](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L44-L68)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L44-L68).

**Why it matters**: `W7DensitySolver.solve()` already scales the density to `mesh.time_explosion` with the homologous-expansion factor \(t^{-3}\). The existing configuration path then applies `calculate_density_after_time()` again, potentially multiplying W7 densities by \((t_\mathrm{explosion}/t_0)^{-3}\) twice. This directly affects mass density, electron density, ionization, opacity, and the emergent spectrum.

**Evidence**: The new tests exercise the solver directly, while regression comparison runs were skipped for this proposed version; the full configuration path therefore lacks equivalent scientific validation.

### Velocity-cell sampling changed by the temporary mesh — close review

**Where to look**: [[temporary mesh construction](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L49-L60)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/parse_density_configuration.py#L49-L60) and [[solver midpoint evaluation](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a7/tardis/io/model/classic/parse_density.py#L94-L97)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a7/tardis/io/model/classic/parse_density.py#L94-L97).

**Why it matters**: The parser receives `v_middle`, which already represents shell centers, but the change treats those values as interfaces by prepending zero. The solver consequently evaluates the first density at \(v_\mathrm{middle}/2\) and later densities at averages of adjacent shell centers, rather than at the original shell centers. For power-law, exponential, and W7 profiles, this changes the density shape and normalization.

**Question for the reviewer**: Is `v_middle` intentionally being reinterpreted as interface velocities here, and does the resulting sampling preserve the intended shell-centered density definition?

### W7 model meaning and validation — check carefully

**Where to look**: [[W7 equation](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L148-L163)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a47228a7/tardis/io/model/classic/parse_density.py#L148-L163) and [[W7 parser selection](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a7/tardis/io/model/classic/parse_density.py#L236-L242)](https://github.com/tardis-sn/tardis/blob/5a420c540f44bed66c98d3561258b298a7/tardis/io/model/classic/parse_density.py#L236-L242).

**Why it matters**: The implementation identifies Branch85 W7 with a pure \(v^{-7}\) law, while the configuration schema describes Branch85 W7 as a seventh-order fit to the W7 model. A specialist should confirm that the intended approximation is indeed the power law and that the normalization at `w7_time_0` and `w7_v_0` reproduces the established W7 density profile.

**Evidence**: The parser test still expects a `PowerLawDensitySolver`, although the proposed parser returns `W7DensitySolver`, and the new W7 tests mostly assert positivity and broad magnitude bounds rather than agreement with an independent W7 reference.

**Version reviewed**: [[PR 3365](https://github.com/tardis-sn/tardis/pull/3365)](https://github.com/tardis-sn/tardis/pull/3365), base `a1a284343ed3cba7acf5cd6606c0ec0e166b8e0c`, proposed `5a420c540f44bed66c98d3561258b298a47228a7`. The proposed code was inspected but not run locally.

File: .agents/skills/tardis-pr-review/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  level and transition mappings, units, missing values, filters, joins, and
+  conversions between tabular and array representations.
+- Observable construction: spectral binning, luminosity integration, virtual or
+  formal-integral spectra, escaping-packet selection, and output quantity
+  definitions.
+- Scientific validation: analytic or limiting cases, independent comparisons,
+  regression baselines, tolerances, packet counts, stochastic variance, test
+  oracle provenance, and assertions on physically meaningful quantities.
+
+Label an area `close review` when the PR changes a physical model, conservation
```
Comment (connor-mcclellan) [inline .agents/skills/tardis-pr-review/SKILL.md]: These labels aren't distinct enough. `close review` and `check carefully` don't actually differentiate between the cases described here.

`Close review` should be fragmented further into more descriptive labels: `Physical model change`, `Convention change`, `Coupled calculation with downstream effects`, `Change in interpretation of scientific data`, `Change in stored reference result`.

`Check carefully` should be split into `Calculation details changed`, `Defaults changed`, `Boundary values changed`, `Tests changed`.

This is just additional text at the top of the item, so there's no reason to collapse it into two arbitrary, less-descriptive labels.

**Review by connor-mcclellan — CHANGES_REQUESTED** [review]

**Comment by andrewfullard** [comment]
Benchmark failures are currently expected

**Review by connor-mcclellan — APPROVED** [review]

**Review by jvshields — APPROVED** [review]
